File: org.boris.xlloop/servers/ruby/xlloop.rb
require "socket"

#  Defines the XLoper types
XL_TYPE_NUM = 1
XL_TYPE_STR = 2
XL_TYPE_BOOL = 3
XL_TYPE_ERR = 4
XL_TYPE_MULTI = 5
XL_TYPE_MISSING = 6
XL_TYPE_NIL = 7
XL_TYPE_INT = 8
XL_TYPE_SREF = 9

# Defines XLError types
XL_ERROR_NULL = 0
XL_ERROR_DIV0 = 7
XL_ERROR_VALUE = 15
XL_ERROR_REF = 23
XL_ERROR_NAME = 29
XL_ERROR_NUM = 36
XL_ERROR_NA = 42

class XLError
  attr_reader :err
  def initialize(err)
    @err = err
  end
  
  def stream_xloper(io)
    io.putc(XL_TYPE_ERR)
    XLCodec.encodeInt(@err, io)
  end
end

class XLSRef
  attr_reader :col_first, :col_last, :rw_first, :rw_last
  def initialize(col_first, col_last, rw_first, rw_last)
    @col_first = col_first
    @col_last = col_last
    @rw_first = rw_first
    @rw_last = rw_last
  end
  
  def stream_xloper(io)
    io.putc(XL_TYPE_SREF)
    XLCodec.encodeInt(@col_first, io)
    XLCodec.encodeInt(@col_last, io)
    XLCodec.encodeInt(@rw_first, io)
    XLCodec.encodeInt(@rw_last, io)
  end
end

class Object
  def stream_xloper(io)
    self.to_s.stream_xloper(io)
  end
end

class String
  def stream_xloper(io)
    io.putc(XL_TYPE_STR)
    if self.length > 255 
      io.putc(255)
      for i in 1..255 
        io.putc(self[i-1])
      end
    else
      io.putc(self.length & 0xff)
      if self.length > 0
        io.write(self)
      end
    end
  end
end

class Numeric
  def stream_xloper(io)
    io.putc(XL_TYPE_NUM)
    io.write([self].pack('G'))
  end
end

class Boolean
  def stream_xloper(io)
    io.putc(XL_TYPE_BOOL)
    io.putc(self ? 1 : 0)
  end
end

class TrueClass
  def stream_xloper(io)
    io.putc(XL_TYPE_BOOL)
    io.putc(1)
  end
end

class FalseClass
  def stream_xloper(io)
    io.putc(XL_TYPE_BOOL)
    io.putc(0)
  end
end

class NilClass
  def stream_xloper(io)
    io.putc(XL_TYPE_NIL)
  end
end

class Array
  def stream_xloper(io)
    io.putc(XL_TYPE_MULTI)
    XLCodec.encodeInt(self.length, io) # rows
    if self.length > 0 
      e = self[0]
      if e.kind_of?(Array) 
        XLCodec.encodeInt(e.length, io) # cols
        for i in 1..self.length
          elem = self[i-1]
          if elem.kind_of?(Array) 
            if elem.length == e.length 
              elem.each { |ee| ee.stream_xloper(io) }
            elsif elem.length < e.length 
              elem.each { |ee| ee.stream_xloper(io) }
              for j in elem.length..(e.length-1)
                io.putc(XL_TYPE_MISSING)
              end
            else
              for j in 1..e.length
                elem[j-1].stream_xloper(io)
              end
            end
          else
            elem.stream_xloper(io)
             (e.length-1).times do
              io.putc(XL_TYPE_MISSING)
            end
          end
        end
      else
        XLCodec.encodeInt(1, io)
        self.each { |elem| elem.stream_xloper(io) }
      end
    else
      XLCodec.encodeInt(0, io)
    end
  end
end

class Hash
  def stream_xloper(io)
    a = Array.new
    self.each { |k,v| 
      b = Array.new
      b.push(k)
      b.push(v)
      a.push(b)
    }
    a.stream_xloper(io)
  end
end

class FunctionContext
  attr_reader :caller, :sheet_name
  def initialize(caller, sheet_name)
    @caller = caller
    @sheet_name = sheet_name
  end
end

class FunctionInformation
  attr_reader :name, :help, :category, :shortcut, :topic
  attr_writer :help, :category, :shortcut, :topic
  
  def initialize(name, help=nil)
    @name = name
    @args = Array.new
    @argHelps = Array.new
    @help = help
    @category = nil
    @shortcut = nil
    @topic = nil
    @isVolatile = false
  end
  
  def addArg(name, help)
    @args.push(name)
    @argHelps.push(help)
  end
  
  def stream_xloper(io)
    h = Hash.new
    h["functionName"] = @name
    h["functionHelp"] = @help if @help != nil
    h["category"] = @category if @category != nil
    h["shortcutText"] = @shortcut if @shortcut != nil
    h["helpTopic"] = @topic if @topic != nil
    h["isVolatile"] = @isVolatile if @isVolatile
    if @args.length > 0
      h["argumentText"] = @args.join(",")
      h["argumentHelp"] = @argHelps
    end
    h.stream_xloper(io)
  end
end

class ReflectionHandler
  def initialize
    o = Object.new
    @baseMethods = [o.public_methods, o.singleton_methods].flatten
    @methods = Hash.new
  end
  
  def push(object, namespace="")
    methods = [object.public_methods, object.singleton_methods].flatten - @baseMethods
    methods.each { |s| 
      m = object.method(s)
      k = namespace + s
      v = @methods[k]
      if v == nil
        @methods[k] = m
      elsif v.kind_of?(Array)
        v.push(m)
      else
        a = Array.new
        a.push(v)
        a.push(m)
        @methods[k] = a
      end
    }
  end
  
  def invoke(context, name, args)
    m = @methods[name]
    chomp(args) # remove trailing nulls
    if m == nil
      if name == "org.boris.xlloop.GetFunctions"
        fi = Array.new
        @methods.each_key { |k| fi.push(FunctionInformation.new(k)) }
        return fi
      else
        return "#Unknown Function"
      end
    elsif m.kind_of?(Array)
      m.each { |e| 
        if canInvoke(e, args.length)
          return invokeM(e, args)
        end
      }
      return "#Invalid number of arguments to #{e}"
    else
      if canInvoke(m, args.length)
        return invokeM(m, args)
      else 
        return "#Invalid number of arguments to #{m}"
      end
    end
  end
  
  protected
  def invokeM(method, args)
    begin
      method.call(*args)
    rescue
      return "#" + $!
    end
  end
  
  def canInvoke(method, argc)
    return (method.arity >= 0 && argc == method.arity) || 
     (method.arity < 0 && (method.arity + argc) >= -1)
  end
  
  def chomp(args)
    while args != nil && args.length > 0 && args.last == nil 
      args.pop
    end
  end
end

class XLCodec
  def XLCodec.decode(io)
    type = io.getc
    case type
      when XL_TYPE_NUM then return io.read(8).unpack('G')[0]
      when XL_TYPE_STR then return io.read(io.getc) 
      when XL_TYPE_BOOL then return io.getc == 0 ? false : true
      when XL_TYPE_ERR then return XLError.new(decodeInt(io))
      when XL_TYPE_MULTI then return decodeMulti(io)
      when XL_TYPE_MISSING then return nil
      when XL_TYPE_NIL then return nil
      when XL_TYPE_INT then return decodeInt(io)
      when XL_TYPE_SREF then return XLSRef.new(decodeInt(io), decodeInt(io), decodeInt(io), decodeInt(io))
    else raise "Invalid XLoper type"
    end
  end
  
  def XLCodec.decodeInt(io)
    io.getc << 24 | io.getc << 16 | io.getc << 8 | io.getc
  end
  
  def XLCodec.decodeMulti(io)
    rows = decodeInt(io)
    cols = decodeInt(io)
    arr = Array.new
    if cols > 1
      rows.times do
        arr2 = Array.new
        cols.times do 
          arr2.push(decode(io))
        end
        arr.push(arr2)
      end
    else
      rows.times do  
        arr.push(decode(io))
      end
    end
    return arr
  end
  
  def XLCodec.encodeInt(value, io)
    io.putc(value >> 24 & 0xff)
    io.putc(value >> 16 & 0xff)
    io.putc(value >> 8 & 0xff)
    io.putc(value & 0xff)
  end
end

class XLLoopServer
  def initialize(handler, port=5454)
    @handler = handler
    @port = port
  end
  
  def start
    @socket = TCPServer.new('localhost', @port)
    while !@socket.closed?
      Thread.start(@socket.accept) { |socket|
        while !socket.closed?
          begin
            context = nil
            name = XLCodec.decode(socket)
            if(!name.kind_of?(String))
              version = name
              if(version == 20)
                extra_info = XLCodec.decode(socket)
                if(extra_info)
                  caller = XLCodec.decode(socket)
                  sheet_name = XLCodec.decode(socket)
                  context = FunctionContext.new(caller, sheet_name)
                end
              else
                "#Unknown protocol version".stream_xloper(socket)
                socket.close
              end
              name = XLCodec.decode(socket)
            end
            argc = XLCodec.decode(socket)
            args = Array.new
            argc.times do
              args.push(XLCodec.decode(socket))
            end
            res = @handler.invoke(context, name, args)
            res.stream_xloper(socket)
            socket.flush
          rescue
            print "Error. ", $!, "\n"
            socket.close
          end
        end
      }
    end
  end
  
  def stop
    @socket.close
  end
end

File: org.boris.xlloop/servers/python/xlloop.py
import random, socket, select, struct, threading, traceback, types, SocketServer
import inspect

#  Defines the XLoper types
XL_TYPE_NUM = 1
XL_TYPE_STR = 2
XL_TYPE_BOOL = 3
XL_TYPE_ERR = 4
XL_TYPE_MULTI = 5
XL_TYPE_MISSING = 6
XL_TYPE_NIL = 7
XL_TYPE_INT = 8
XL_TYPE_SREF = 9

# Defines XLError types
XL_ERROR_NULL = 0
XL_ERROR_DIV0 = 7
XL_ERROR_VALUE = 15
XL_ERROR_REF = 23
XL_ERROR_NAME = 29
XL_ERROR_NUM = 36
XL_ERROR_NA = 42

class XLError:
    def __init__(self, err):
        self.err = err

class XLSRef:
    def __init__(self, col_first, col_last, rw_first, rw_last):
        self.col_first = col_first
        self.col_last = col_last
        self.rw_first = rw_first
        self.rw_last = rw_last

class XLCodec:
    def decode(socket):
        type = ord(socket.recv(1))
        if type == XL_TYPE_NUM:
            return struct.unpack('>d', socket.recv(8))[0]
        elif type == XL_TYPE_STR:
            len = ord(socket.recv(1))
            if len == 0:
                return ''
            return socket.recv(len)
        elif type == XL_TYPE_BOOL:
            if ord(socket.recv(1)) == 0:
                return False
            else:
                return True
        elif type == XL_TYPE_ERR:
            return XLError(XLCodec.decodeInt(socket))
        elif type == XL_TYPE_MULTI:
            rows = XLCodec.decodeInt(socket)
            cols = XLCodec.decodeInt(socket)
            if cols == 0 or rows == 0:
                return []
            a = []
            if cols > 1:
                for i in xrange(rows):
                    aa = []
                    for j in xrange(cols):
                        aa.append(XLCodec.decode(socket))
                    a.append(aa)
            else:
                for i in xrange(rows):
                    a.append(XLCodec.decode(socket))
            return a
        elif type == XL_TYPE_MISSING:
            return None
        elif type == XL_TYPE_NIL:
            return None
        elif type == XL_TYPE_INT:
            return XLCodec.decodeInt(socket)
        elif type == XL_TYPE_SREF:
            return XLSRef(XLCodec.decodeInt(socket), XLCodec.decodeInt(socket), XLCodec.decodeInt(socket), XLCodec.decodeInt(socket))
        else:
            raise TypeError("Invalid XLoper type encountered")
    decode = staticmethod(decode)

    def decodeInt(socket):
        l = socket.recv(4)
        return ord(l[0]) << 24 | ord(l[1]) << 16 | ord(l[2]) << 8 | ord(l[3])
    decodeInt = staticmethod(decodeInt)

    def encode(value, socket):
        if isinstance(value, types.StringType):
            socket.send(struct.pack('B', XL_TYPE_STR))
            l = len(value)
            socket.send(struct.pack('B', l))
            if l > 0:
                socket.send(value)
        elif isinstance(value, types.FloatType):
            socket.send(struct.pack('B', XL_TYPE_NUM))
            socket.send(struct.pack('>d', value))
        elif isinstance(value, types.NoneType):
            socket.send(struct.pack('B', XL_TYPE_NIL))
        elif isinstance(value, XLError):
            socket.send(struct.pack('B', XL_TYPE_ERR))
            socket.send(struct.pack('>i', value.err))
        elif isinstance(value, XLSRef):
            socket.send(struct.pack('B', XL_TYPE_SREF))
            socket.send(struct.pack('>i', value.col_first))
            socket.send(struct.pack('>i', value.col_last))
            socket.send(struct.pack('>i', value.rw_first))
            socket.send(struct.pack('>i', value.rw_last))
        elif isinstance(value, types.BooleanType):
            socket.send(struct.pack('B', XL_TYPE_BOOL))
            if(value == True):
                socket.send(struct.pack('B', 1))
            else:
                socket.send(struct.pack('B', 0))
        elif isinstance(value, types.IntType):
            socket.send(struct.pack('B', XL_TYPE_NUM))
            socket.send(struct.pack('>d', float(value)))
        elif isinstance(value, types.ListType):
            socket.send(struct.pack('B', XL_TYPE_MULTI))
            rows = len(value)
            socket.send(struct.pack('>i', rows))
            if rows == 0:
                socket.send(struct.pack('>i', 0)) # zero cols
            else:
                v = value[0]
                if isinstance(v, types.ListType):
                    cols = len(v)
                    socket.send(struct.pack('>i', cols))
                    for i in xrange(rows):
                        v = value[i]
                        if isinstance(v, types.ListType):
                            l = len(v)
                            if l < cols:
                                for j in xrange(l):
                                    XLCodec.encode(v[j], socket)
                                for j in xrange(l, cols):
                                    XLCodec.encode(None, socket)
                            else:
                                for j in xrange(cols):
                                    XLCodec.encode(v[j], socket)
                        else:
                            XLCodec.encode(v, socket)
                            for j in xrange(1, cols):
                                XLCodec.encode(None, socket)
                else:
                    socket.send(struct.pack('>i', 1))
                    for i in xrange(rows):
                        XLCodec.encode(value[i], socket)
        elif isinstance(value, types.TupleType):
            socket.send(struct.pack('B', XL_TYPE_MULTI))
            rows = len(value)
            socket.send(struct.pack('>i', rows))
            if rows == 0:
                socket.send(struct.pack('>i', 0))
            else:
                socket.send(struct.pack('>i', 1)) # only one column for tuples
            for i in xrange(rows):
                XLCodec.encode(value[i], socket)
        else:
            XLCodec.encode(str(value), socket)
    encode = staticmethod(encode)

class FunctionContext:
    def __init__(self, caller, sheet_name):
        self.caller = caller
        self.sheet_name = sheet_name

class XLLoopHandler(SocketServer.BaseRequestHandler):
    def handle(self):
        while 1:
            try:
                context = None
                name = XLCodec.decode(self.request)
                if isinstance(name, types.IntType):
                    version = name
                    if version == 20:
                        extra_info = XLCodec.decode(self.request)
                        if extra_info:
                            caller = XLCodec.decode(self.request)
                            sheet_name = XLCodec.decode(self.request)
                            context = FunctionContext(caller, sheet_name)
                    else:
                        raise TypeError("Unknown protocol version")
                    name = XLCodec.decode(self.request)
                argc = XLCodec.decode(self.request)
                args = []
                for i in xrange(argc):
                    args.append(XLCodec.decode(self.request))
                res = self.server.handler.invoke(context, name, args)
                XLCodec.encode(res, self.request)
            except:
                traceback.print_exc()
                self.request.shutdown
                break

class XLLoopServer:
    def __init__(self, handler, port=5454):
        self.handler = handler
        self.port = port

    def start(self):
        self.server = SocketServer.ThreadingTCPServer(('localhost', self.port), XLLoopHandler)
        self.server.handler = self.handler
        self.server.serve_forever()

    def stop(self):
        self.server.shutdown()

class FnServer(object):
    def __init__(self):
        self.exposed_fns = {}
        self._get_fns = []

    def add_fn(self, excel_name, fn_to_call, doc=None, arg_descrs=[], args='', register=True):
        self.exposed_fns[excel_name] = fn_to_call
        if register:
            excel_info = [
                    ["functionName", excel_name],
                    ["functionHelp", doc],
                    ["category", "PYTHON"],
                    ["argumentHelp", arg_descrs],
                    ["argumentText", args]
                ]
            self._get_fns.append(excel_info)

    def invoke(self, context, name, args):
        if name in self.exposed_fns:
            try:
                return self.exposed_fns[name](*args)
            except TypeError, msg:
                print msg
                return
        else:
            return XLError(XL_ERROR_NAME)

    def get_fns(self):
        return tuple(self._get_fns,)

fn_server = FnServer()
fn_server.add_fn("org.boris.xlloop.GetFunctions", fn_server.get_fns, register=False)

def expose(excel_name=None, arg_descrs=[]):
    def decorator(fn):
        xln = excel_name # Need local reference, otherwise get exception
        if xln is None:
            xln = fn.__name__
        d = fn.func_doc
        args = ", ".join([x.title() for x in inspect.getargspec(fn)[0]])
        argd = arg_descrs
        fn_server.add_fn(xln, fn, doc=d, arg_descrs=argd, args=args)
        return fn
    return decorator

def run(port=5454):
    print 'Starting Function Server on port %d...' %port
    xs = XLLoopServer(fn_server, port)
    xs.start()
